### PR TITLE
Prevent data duplication on unspill

### LINKED ISSUE #3756: Spill to disk may cause data duplication
In aggressive spill-to-disk scenarios I observed that distributed may spill all the data it has in memory while still complaining with the following message that there is no more data to spill

> Memory use is high but worker has no data "
                        "to store to disk.  Perhaps some other process "
                        "is leaking memory?
Side note: In our setup every worker runs in an isolated container so the chance of another process interfering with it is virtually zero.

It is true that these workers do indeed still hold on to a lot of memory without it being very transparent about where the data is. GC is useless, i.e. it is **not** related to https://github.com/dask/zict/issues/19

Investigating this issue let me realise that we may try to spill data which is actually currently in use.
The most common example for this is probably once the worker collected the dependencies and schedules the execution of the task, see [here](https://github.com/dask/distributed/blob/c8ee44bebe40e4db4495be1e2e0340fafbb41a37/distributed/worker.py#L2485). Another one would be when data is [requested form a worker](https://github.com/dask/distributed/blob/c8ee44bebe40e4db4495be1e2e0340fafbb41a37/distributed/worker.py#L1232) and we're still serializing/submitting it.
I could _prove_ this assumption by patching the worker code, tracking the keys in-execution/in-spilling with some logging and it turns out that for some jobs I hold on to multiple GBs of memory although it was supposedly already spilled.

If we spill data which is currently still in use, this is not only misleading to the user, since the data is still in memory, but it may also cause heavy data duplication if the spilled-but-still-in-memory dependency is requested by another worker since the original data is still in memory but the buffer would fetch the key from the slow store and materialise it a second time since it doesn't know of the original object anymore. If this piece of data is requested by more than one worker, the data could be duplicated multiple times even.

In non-aggressive spill-to-disk scenarios we should be protected from this by the LRU in the buffer but if memory pressure is high, the spilling might actually worsen the situation in these cases.

My practical approach to this would be to introduce something like a `(un)lock_key_in_store` method to the buffer which protects the key from spilling and manually (un)setting this in the distributed code. If there is a smarter approach, I'd be glad to hear about it.

Also, If my reasoning is somewhere flawed, I'd appreciate feedback since, so far, I could only prove that we try to spill data currently in use but, so far, the duplication is just theory.

Related issues:
https://github.com/dask/dask/issues/2456

### PR DESCRIPTION
**Pull request by crusaderky**
- Closes #3756
- Blocked by https://github.com/dask/zict/pull/65

### CODE REVIEW

**Review by fjetter — COMMENTED** [review]
Very nice! Looking forward to having this in! Changes LGTM once zict is merged, etc.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -47,4 +47,6 @@ repos:
           - numpy
           - dask
           - tornado
-          - zict
+          # TEMP revert before merging
+          # - zict
+          - git+https://github.com/crusaderky/zict@cache
```
Comment (crusaderky) [inline .pre-commit-config.yaml:52]: TODO revert

File: continuous_integration/environment-3.9.yaml [hunk 2]
```diff
@@ -52,7 +52,9 @@ dependencies:
   - pip:
       - git+https://github.com/dask/dask
       - git+https://github.com/dask/s3fs
-      - git+https://github.com/dask/zict
+      # TEMP revert before merging
+      # - git+https://github.com/dask/zict
+      - git+https://github.com/crusaderky/zict@cache
```
Comment (crusaderky) [inline continuous_integration/environment-3.9.yaml:57]: TODO revert

File: distributed/tests/test_spill.py [hunk 3]
```diff
@@ -349,7 +282,63 @@ def test_spillbuffer_evict(tmpdir):
 
     assert "Failed to pickle" in logs_evict_key.getvalue()
     # bad keys stays in fast
-    assert set(buf.fast) == {"a_bad"}
-    assert buf.fast.weights == {"a_bad": sizeof(a_bad)}
-    assert set(buf.slow) == {"a"}
-    assert buf.slow.weight_by_key == {"a": psize(a)}
+    assert_buf(buf, {"bad": bad}, {"a": a})
+
+
```
Comment (crusaderky) [inline distributed/tests/test_spill.py:287]: new test start here; everything else in this module is just a refactor

**Pull request merged**